driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -93,56 +86,78 @@ impl StableXOrderBookReading for UpdatingOrderbook {
+        let to_block = BlockNumber::Number(current_block);
         select_biased! {
             _ = exit_indicator => return Ok(()),
-            event = stream.next() => {
-                log::info!("Received new event.");
-                let event = event.ok_or(anyhow!("stream ended"))??;
-                handle_event(&orderbook, &mut block_timestamp_reader, event).await?;
-            },
-            past_events = past_events => {
-                let past_events = past_events?;
-                log::info!("Received {} past events.", past_events.len());
-                let block_hashes = past_events.iter().map(|event| {
-                    let metadata = event.meta.as_ref().ok_or(anyhow!("event without metadata: {:?}", event))?;
-                    Ok(metadata.block_hash)
-                }).collect::<Result<HashSet<H256>>>()?;
-                block_timestamp_reader.prepare_cache(block_hashes).await?;
-                for event in past_events {
-                    handle_event(&orderbook, &mut block_timestamp_reader, event).await?;
-                }
-                log::info!("Finished applying past events");
-                orderbook_ready.store(true, Ordering::SeqCst);
+            events = contract.past_events(BlockNumber::Number(from_block.into()), to_block).fuse() => {
+                handle_events(&orderbook, &mut block_timestamp_reader, events?, from_block).await?;
             },
         };
+        last_handled_block = current_block.as_u64();
+        // This will be set and stay true the first time we reach this point.
+        orderbook_ready.store(true, Ordering::SeqCst);
+        // TODO: This is not optimal because it means we might sleep for some time when exit
+        // indicator triggered. We should create a sleep future that expires after poll interval and
+        // join both together.
+        std::thread::sleep(POLL_INTERVALL);
     }
 }
 
+/// Apply a vector of events to the orderbook.
+async fn handle_events(
+    orderbook: &Mutex<Orderbook>,
+    block_timestamp_reader: &mut CachedBlockTimestampReader<Web3>,
+    events: Vec<Event<batch_exchange::Event>>,
+    delete_events_starting_at_block: u64,
+) -> Result<()> {
+    log::info!("Received {} events.", events.len());
+    let block_hashes = events
+        .iter()
+        .map(|event| {
+            let metadata = event
+                .meta
+                .as_ref()
+                .ok_or_else(|| anyhow!("event without metadata: {:?}", event))?;
+            Ok(metadata.block_hash)
+        })
+        .collect::<Result<HashSet<H256>>>()?;
+    block_timestamp_reader.prepare_cache(block_hashes).await?;
+    // Locking here ensures that the orderbook is not observable after the events have been deleted
+    // but the new events not yet applied.
+    let mut orderbook = orderbook
+        .lock()
+        .map_err(|err| anyhow!("poison error: {}", err))?;
+    orderbook.delete_events_starting_at_block(delete_events_starting_at_block);
```

Comment:
I find it awkward that we are calling a function whose name is the same as the variable we pass it.

Also, why are we "deleting events"?

Reply:
Because we are refetching all events from `delete_events_starting_at_block` and thus we want to delete our prior version of these events (as they might have changed in a reorg).

For naming I'd suggest something like `refetch_cutoff_block` or the like, but I don't think it's important enough right now to justify another round trip of travis.